### PR TITLE
Remove fragment state vector on CacheChange_t [6623]

### PR DESCRIPTION
**Pull request by MiguelCompany**
This is part of the Fragmentation API refactor.

The `dataFragments_` vector is removed by having the list of missing fragments implemented inside the data payload. The first 4 bytes of each fragment will have the index of the next missing fragment.

### CODE REVIEW

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/697/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/596/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/692/)
* Performance [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Performance%20Tests%20Localhost/162/)

File: include/fastdds/rtps/common/CacheChange.h [hunk 1]
```diff
@@ -149,22 +152,26 @@ struct RTPS_DllAPI CacheChange_t
...
+            dataFragments_->assign(ch_ptr->dataFragments_->begin(), ch_ptr->dataFragments_->end());
+        }
 
         isRead = ch_ptr->isRead;
     }
 
     ~CacheChange_t()
     {
-        if (dataFragments_)
+        if (dataFragments_ != nullptr)
```
Comment (LuisGP) [inline include/fastdds/rtps/common/CacheChange.h]: This check is unnecessary, `delete nullptr;` is a non-op.

File: src/cpp/rtps/reader/FragmentedChangePitStop.cpp [hunk 2]
```diff
@@ -69,51 +69,17 @@ CacheChange_t* FragmentedChangePitStop::process(
...
-            }
-
-            original_change_cit->getChange()->getDataFragments()->at(count) = ChangeFragmentStatus_t::PRESENT;
+    size_t original_offset = size_t(fragment_starting_num - 1) * original_change_cit->getChange()->getFragmentSize();
 
-            was_updated = true;
-        }
-    }
+    memcpy(
+        &original_change_cit->getChange()->serializedPayload.data[original_offset],
```
Comment (LuisGP) [inline src/cpp/rtps/reader/FragmentedChangePitStop.cpp]: Oh! Nice catch :D

File: include/fastdds/rtps/common/CacheChange.h [hunk 3]
```diff
@@ -247,9 +247,17 @@ struct RTPS_DllAPI CacheChange_t
...
+            uint32_t last_fragment = initial_fragment + num_of_fragments;
+            if (initial_fragment <= first_missing_fragment_)
             {
-                first_missing_fragment_ += num_of_fragments;
+                // Perform first = *first until first >= last_received
+                while (first_missing_fragment_ < last_fragment)
+                {
+                    size_t offset = fragment_size_;
+                    offset *= first_missing_fragment_;
+                    uint32_t* fragment = (uint32_t*)&serializedPayload.data[offset];
```
Comment (LuisGP) [inline include/fastdds/rtps/common/CacheChange.h]: Try to adopt the habit to use C++11 castings.

File: include/fastdds/rtps/common/CacheChange.h [hunk 4]
```diff
@@ -306,6 +295,30 @@ struct RTPS_DllAPI CacheChange_t
...
+        *ptr = next_fragment_index;
+    }
+
+    uint32_t* next_fragment_pointer(
+            uint32_t fragment_index)
+    {
+        size_t offset = fragment_size_;
+        offset *= fragment_index;
+        offset = (offset + 3) & ~3;
+        return (uint32_t*)(&serializedPayload.data[offset]);
```
Comment (LuisGP) [inline include/fastdds/rtps/common/CacheChange.h]: Use C++11 castings.

File: include/fastdds/rtps/common/CacheChange.h (hunk 1 above)
Comment (MiguelCompany) [reply include/fastdds/rtps/common/CacheChange.h]: dataFragments_ has been removed

File: include/fastdds/rtps/common/CacheChange.h (hunk 3 above)
Comment (MiguelCompany) [reply include/fastdds/rtps/common/CacheChange.h]: done

File: include/fastdds/rtps/common/CacheChange.h (hunk 4 above)
Comment (MiguelCompany) [reply include/fastdds/rtps/common/CacheChange.h]: done

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/711/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/604/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/706/)
* Performance [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Performance%20Tests%20Localhost/173/)

**Comment by raquelalvarezbanos** [comment]
Merging, as the last commit has minimal changes and tests had passed in the previous builds.

**Pull request merged**